go/pserver/cclient/cclient.go

```diff
@@ -100,9 +105,39 @@ func paddle_new_pserver_client(addrs *C.char, selected int) C.paddle_pserver_cli
 }
 
 //export paddle_new_etcd_pserver_client
-func paddle_new_etcd_pserver_client(etcd_addr *C.char) C.paddle_pserver_client {
-	// TODO(helin): fault tolerant pserver client using etcd.
-	panic("not implemented.")
+func paddle_new_etcd_pserver_client(etcd_addr *C.char, pserver_path *C.char, selected int) C.paddle_pserver_client {
+	if etcd_addr == "" {
+		return C.PSERVER_ERROR
+	}
+	etcd_addrs := strings.Split(etcd_addr, ",")
+	cli, err := clientv3.New(clientv3.Config{
+		Endpoints:   etcd_addrs,
+		DialTimeout: etcdTimeout,
+	})
+	if err != nil {
+		log.Errorln(err)
+		return err
+	}
+	retryTimes := defaultRetryTimes
+	for retryTimes < 0 {
+		retryTimes--
+		ctx, cancel := context.WithTimeout(context.Background(), time.Second)
+		resp, err := cli.Get(ctx, pserver_path)
+		cancel()
+		if err != nil {
+			log.Errorf("get %s error %v", pserver_path, err)
+		}
+		kvs := resp.Kvs
+		if len(kvs) == 0 {
+			log.Infoln("Waiting for pservers register, sleeping")
+			time.Sleep(time.Second)
+			continue
+		}
+		ps_addr := string(kvs[0].Value)
```

Comment:
From design doc:  https://github.com/PaddlePaddle/Paddle/tree/develop/doc/design/cluster_train#trainer-process-1

`ps_addr` is not fetched from one etcd key. Addresses are read from `/ps/<id>` which stands for IP address for each pserver, so `ps_addr` need to concat all these addresses.

And each trainer must "watch" pserver nodes hand handle events of pserver creation and loss.